cmd/zc_traverser_list.go

```diff
@@ -0,0 +1,127 @@
+		}
+	}
+
+	// in case the list of entities was not read properly, we can no longer continue enumeration
+	if err := scanner.Err(); err != nil {
+		return fmt.Errorf("unable to scan %s to retrieve the required list of entities to transfer", l.fileLocation)
+	}
+
+	return nil
+}
+
+func (l *listTraverser) setPreProcessingCallback(callback preProcessingCallback) {
+	// TODO cannot fulfill this request since listTraverser is a meta-traverser and does not
+}
+
+func newListTraverser(parent string, parentSAS string, credential common.CredentialInfo, recursive bool,
+	fileLocation string, parentType common.Location) resourceTraverser {
+	var traverserGenerator childTraverserGenerator
+
+	// when scanning a child path under the parent, we need to make sure that the relative paths of
+	// the results are indeed starting right under the parent
+	// ex: parent = /usr/foo
+	// case 1: child1 is a file
+	//         the relative path returned by the child traverser would be ""
+	//         it should be "child1" instead
+	// case 2: child2 is a directory, and it has items under it such as child2/grandchild1
+	//         the relative path returned by the child traverser would be "grandchild1"
+	//         it should be "child2/grandchild1" instead
+	getPreProcessingCallback := func(relativeChildPath string) preProcessingCallback {
+		return func(object storedObject) storedObject {
+			if object.relativePath == "" {
+				// if the child is a single file
+				object.relativePath = relativeChildPath
+			} else {
+				object.relativePath = common.GenerateFullPath(relativeChildPath, object.relativePath)
+			}
+
+			return object
+		}
+	}
+
+	// TODO implement for Local, ADLS Gen 2
+	switch parentType {
+	case common.ELocation.Blob():
+		traverserGenerator = func(relativeChildPath string) (resourceTraverser, error) {
+			// assume child path is not URL-encoded yet, this is consistent with the behavior of previous implementation
+			childURL, _ := url.Parse(parent)
+			childURL.Path = common.GenerateFullPath(childURL.Path, relativeChildPath)
+
+			// construct traverser that goes through child
+			traverser, err := newBlobTraverserForCopy(childURL.String(), parentSAS, credential, recursive)
+			if err != nil {
+				return nil, err
+			}
+
+			// make sure the returned storedObjects have the right relative path starting at under parent
+			traverser.setPreProcessingCallback(getPreProcessingCallback(relativeChildPath))
+			return traverser, nil
+		}
+	case common.ELocation.File():
+		traverserGenerator = func(relativeChildPath string) (resourceTraverser, error) {
+			// assume child path is not URL-encoded yet, this is consistent with the behavior of previous implementation
+			childURL, _ := url.Parse(parent)
+			childURL.Path = common.GenerateFullPath(childURL.Path, relativeChildPath)
+
+			// construct traverser that goes through child
+			traverser, err := newFileTraverserForCopy(childURL.String(), parentSAS, credential, recursive)
```

Comment:
Can we stop making these highly situational constructors and just have a _global_ traverser generator? Ask to take a look at my initResourceTraverser function sometime today, since it currently only lives on a local repo since the refactor is still very much a WIP.

Reply:
The reason I'm asking is because all this is doing is creating a lot of pointless code duplication that we're inevitably going to have a pain cleaning up.

---
In reply to: [305028492](https://github.com/Azure/azure-storage-azcopy/pull/509#discussion_r305028492) [](ancestors = 305028492)

Reply:
Yes, we can certainly roll them into one constructor.